Refactor

serpient opened the pull request:
![gif](https://media.giphy.com/media/SHdUoTyvbDLc1XRv4S/giphy.gif)

# Entry to game is `TicTacToe.play()`

# Things refactored
- Changed UpdateBoard to Board
- Move board related functions like checking for empty spaces and checking whether an individual space is empty into Board.
- Rename wins test file to match naming convention
- Couple the no_win tuple with `:ok` instead of `:error`.

serpient committed:
passes wins_result chains to win game message

serpient committed:
remove handle_play; make specific handle_win fn

serpient committed:
add another tie check test

serpient committed:
wins result returns board data

serpient committed:
refactor win and validation result so it can be chained

serpient committed:
move handle_validation_result into ProcessInput

serpient committed:
update handle_validation_result tests

serpient committed:
update analyze fn to take in tuple

serpient committed:
update game output (remove "-")

serpient committed:
remove status tuple in wins result

serpient committed:
pass error from validation down entire play chain

serpient committed:
uncomment test unit -- it can catch errors again

serpient committed:
move handle_win to game output

serpient committed:
move get player input to GameOutput

serpient commented on lib/ttt.ex:
```diff
@@ -34,27 +13,22 @@ defmodule TTT do
 
-  def handle_validation_result(input_validation_result) do
-    {status, results} = input_validation_result
+  def handle_win_check_result(win_result, initial_board, current_player) do
```
all errors from my chained play function are bubbled to this last function.
Is this correct? Or should I try and handle errors earlier in the chain as I hit them.

serpient commented on lib/ttt.ex:
```diff
@@ -80,12 +54,10 @@ defmodule TTT do
+        GameOutput.get_player_input(current_player, prompt)
+        |> ProcessInput.handle_input(board, current_player)
+        |> CheckForWins.analyze()
+        |> handle_win_check_result(board, current_player)
```
Chained play function

Currently, errors that occur within this chain is bubbled to the bottom function to handle. Is this the correct way to do this?

aleph-naught2tog replied:
In this case, I think I would opt to split things up a little more -- bubbling errors *is* very good practice and this is a great start. It does feel weird to pass errors around for sure, and takes some getting used to.

The main reason I'd split things up a little more, here, is only sort of related to the errors. 

The biggest reason is: look at your function names. `get_player_input` hands off to `handle_input` -- that makes _total_ sense to me. Super legit and a super common pattern: get data, handle data. I would 100000% expect those two functions to relate to each other and one to flow into the other.

The "flow" from `handle_input`, however, to `analyze` is harder to follow. I had to go look at the functions themselves -- which isn't a bad thing, but can definitely be a sign that something might be clearer written differently. 

In this case, `analyze` takes a board as its parameter -- that's super reasonable. Our function names even give us a clue here: `get_player_input` and `handle_input` are both about `input`, and then ... the input "disappears" (from the function names). That's another sign maybe to split things apart here.

If we go look inside `ProcessInput.handle_input` we get another good clue -- the first few calls are about `input` again, and the last one is about updating a board ... and `analyze` takes a board as a parameter.

Some folks would probably fight me on this, but -- if you think about functions as just inside-out-methods (`handle_input(input)` vs `input.handle`, `update_board(board)` vs `board.update` or `board.analyze` -- you can get a good idea of how things can flow naturally.

If you split those up a little, you can then pattern match based on the result of `handle_input` -- basically, it could be something like "if we get an `:error` we know something is wrong, and there's no sense analyzing the board. otherwise, we should analyze the board."

serpient reviewed (commented):
Questions -- See above. Any other feedback would also be very welcome!

serpient commented on lib/ttt.ex:
```diff
@@ -34,27 +13,22 @@ defmodule TTT do
-        handle_error(input_validation_result, board, current_player)
+        GameOutput.print_win(updated_board, current_player)
     end
   end
```
all errors from my chained play function are bubbled to this last function.
Is this correct? Or should I try and handle errors earlier in the chain as I hit them.

serpient committed:
remove unnecessary paranthesis

serpient committed:
move check for tie to Wins file; refactor code and test

aleph-naught2tog commented on lib/ttt.ex:
```diff
@@ -1,30 +1,4 @@
 defmodule TTT do
```
This is one of those things that I know folks disagree on but -- you might want to consider changing the module name/file to be `TicTacToe` / `tic_tac_toe.ex`. I know it's sort of a pain to write out but my own (personal) preference when writing/reading code is to avoid _any_ abbreviations; it's sooooo tempting but it can really add a lot of clarity to avoid them, same way really great function and variable names like you are doing everywhere really adds clarity!
```suggestion
defmodule TicTacToe do
```

serpient replied:
Gotcha!! That makes sense and will update 👯‍♂️

aleph-naught2tog commented on lib/check_for_wins.ex:
```diff
@@ -30,18 +30,32 @@ defmodule CheckForWins do
-  def analyze(board) do
-    horizontal = convert_horizontal_to_row(board) |> check_row
+  def has_empty_spaces?(board) do
+    Map.values(board)
```
This is nitpicky (which is good! 😁little things means the big things are being taken care of already 😊) but -- it's usually considered clearer/good practice to if possible only have "bare" values at the top of a pipe.
```suggestion
    board
    |> Map.values()
```

serpient replied:
I see -- makes it easier to see what the chain starts from. Will update!

serpient committed:
rename main file to TicTacToe for clarity

serpient committed:
bare values at top of pipes

serpient committed:
update test handle_validation_result per refactor

serpient committed:
refactor cond / case accordingly

serpient committed:
refactor to extract board update in chain; errors are still being passed through

serpient committed:
rename game output to gameIO; it handles input/output

serpient committed:
rename input_test to process_input_test for clarity

serpient committed:
extract board update fn to its own file

serpient committed:
refactor to only handle input processing; returns a string

serpient committed:
add handle_win_check fn

serpient committed:
update defmodules names

serpient committed:
update defmodule name

serpient committed:
refactor analyze to only take board

serpient committed:
add get_message test units

serpient committed:
add error case for board update fn

serpient committed:
add text configuration

serpient committed:
handle win check returns error tuple instead of call to TTT.handleError

serpient committed:
add handle_win_check test units

serpient committed:
add failing test unit for no_win condition

serpient committed:
update board_is_filled to game_is_tie for clarity

serpient committed:
fix bug in which continue game was not triggering correctly

no_win is technically an "error", so it was triggering a handle_error fn , not the intended continue play fn. Moved error handling to the bottom to fix.

serpient committed:
commented out no-win case check as it triggers a function call -- not sure how to test that it continued

serpient committed:
mix format

serpient committed:
refactor test descriptions for clarity

serpient committed:
add analyze header to test name

serpient committed:
add win methods test units

serpient committed:
update tie message to include final board print

aaizuss reviewed (commented):
I'm going to leave more specific comments, but in general, I think you have modules and functions that do too much (violating the single responsibility principle). Specifically, the `ProcessInput` and `CheckForWins` modules appear to know too much about the way the board is represented internally (for example, they both have logic that checks whether a space/spaces are empty, and they know that the board is a map). You can abstract the concept of a board into another module, so someone reading your code does not have to know exactly how a board is implemented in order to understand how the other functions work. This will also make the board implementation easier to change - if for some reason you decide to represent the board differently, it's going to be challenging to find all the places in your code that depend on the current board implementation until you have some sort of way to interact with the board via functions like mark_space, has_empty_spaces, etc. If you use abstractions like this, changing how the board works is easier because then you won't need to look through all the places in your code that are coupled to a specific implementation.

aaizuss commented on lib/check_for_wins.ex:
```diff
@@ -30,6 +31,12 @@ defmodule CheckForWins do
     end
   end
 
+  def has_empty_spaces?(board) do
```
Why does check for wins know how a board is implemented? And there's another place in your code where you check if spaces are empty - think of where this logic might belong

aaizuss commented on lib/check_for_wins.ex:
```diff
@@ -40,8 +47,18 @@ defmodule CheckForWins do
-      true -> {:error, :no_win}
+      Enum.member?(board_has_win?, true) -> {:ok, :wins_game, board}
+      has_empty_spaces?(board) == false -> {:error, :game_is_a_tie, board}
+      true -> {:error, :no_win, board}
```
why do you return an error tuple when no one wins a game?

aaizuss commented on lib/game_IO.ex:
```diff
@@ -1,10 +1,10 @@
-defmodule GameOutput do
+defmodule GameIO do
   def get_message(key) do
```
I like the way you've organized the possible messages into a map 👍

aaizuss commented on lib/process_input.ex:
```diff
@@ -38,13 +30,9 @@ defmodule ProcessInput do
   end
 
   def handle_input(input, board, current_player) do
```
I like the way this function flows

aaizuss commented on lib/process_input.ex:
```diff
@@ -27,7 +19,7 @@ defmodule ProcessInput do
 
   def validate_input(result, board, _current_player) when is_integer(result) do
```
I think some of the details in this function (is_empty and that it knows how many spaces are on a board) should be defined elsewhere and then used by this function

aaizuss commented on lib/update_board.ex:
```diff
@@ -0,0 +1,10 @@
+defmodule UpdateBoard do
```
Perhaps this module should be renamed to Board and be responsible for basic board functionality (marking a space, checking if a space is taken, etc)

aaizuss commented on test/wins_test.exs:
```diff
@@ -0,0 +1,328 @@
+defmodule Check_For_Wins_Test do
```
You should rename this file so that it is consistent with the implementation file name (`check_for_wins_test.exs`)

serpient committed:
rename wins test file to match naming convention

serpient committed:
rename update board to board; move in board fns

serpient committed:
set no_win to :ok instead of :error

- its not an error, so it wasnt very clear to lump it inside as a error

serpient committed:
move board conersion fns into Board
